src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    private void decompress(Inflater inflater, byte[] data, ByteArrayOutputStream outputBuffer) throws DataFormatException{
+        inflater.setInput(data);
+        byte[] buffer = new byte[BUFFER_SIZE];
+
+        int bytesInflated;
+        while((bytesInflated = inflater.inflate(buffer)) > 0){
+            outputBuffer.write(buffer, 0, bytesInflated);
+        }
+    }
+
+    @Override
+    public void encodeFrame(Framedata inputFrame) {
+        // Only DataFrames can be decompressed.
+        if(!(inputFrame instanceof DataFrame))
+            return;
+
+        // Only the first frame's RSV1 must be set.
+        if(!(inputFrame instanceof ContinuousFrame))
+            ((DataFrame) inputFrame).setRSV1(true);
+
+        Deflater deflater = new Deflater(Deflater.DEFAULT_COMPRESSION, true);
+        deflater.setInput(inputFrame.getPayloadData().array());
+        deflater.finish();
+
+        // Compressed output buffer.
+        ByteArrayOutputStream output = new ByteArrayOutputStream();
+        // Temporary buffer to hold compressed output.
+        byte[] buffer = new byte[1024];
+        int bytesCompressed;
+        while((bytesCompressed = deflater.deflate(buffer)) > 0) {
+            output.write(buffer, 0, bytesCompressed);
+        }
+        deflater.end();
+
+        byte outputBytes[] = output.toByteArray();
+        int outputLength = outputBytes.length;
+        /*
+            https://tools.ietf.org/html/rfc7692#section-7.2.1 states that if the final fragment's compressed
+                payload ends with 0x00 0x00 0xff 0xff, they should be removed.
+            To simulate removal, we just pass 4 bytes less to the new payload
+                if the frame is final and outputBytes ends with 0x00 0x00 0xff 0xff.
+         */
+        if(inputFrame.isFin() && endsWithTail(outputBytes))
+            outputLength -= TAIL_BYTES.length;
+
+        // Set frames payload to the new compressed data.
+        ((FramedataImpl1) inputFrame).setPayload(ByteBuffer.wrap(outputBytes, 0, outputLength));
+    }
+
+    private boolean endsWithTail(byte[] data){
+        if(data.length < 4)
+            return false;
+
+        int length = data.length;
+        for(int i = 0; i <= TAIL_BYTES.length; i--){
+            if(TAIL_BYTES[i] != data[length - TAIL_BYTES.length + i])
+                return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean acceptProvidedExtensionAsServer(String inputExtension) {
+        String[] requestedExtensions = inputExtension.split(",");
+        for(String extension : requestedExtensions)
+            if(EXTENSION_REGISTERED_NAME.equalsIgnoreCase(extension.trim()))
+                return true;
+
+        return false;
+    }
+
+    @Override
+    public boolean acceptProvidedExtensionAsClient(String inputExtension) {
+        String[] requestedExtensions = inputExtension.split(",");
```

Comment:
Should we not split by ; ?

Reply:
I believe [specification ](https://tools.ietf.org/html/rfc6455#section-9) states that extensions are seperated by commas( , ), and parameters to extensions are seperated by semi-colons( ; ).
> 
        Sec-WebSocket-Extensions = extension-list
         extension-list = 1#extension
         extension = extension-token *( ";" extension-param )
         extension-token = registered-token
         registered-token = token
         extension-param = token [ "=" (token | quoted-string) ]
             ;When using the quoted-string syntax variant, the value
             ;after quoted-string unescaping MUST conform to the
             ;'token' ABNF.

An example from the spec:
`Spec-WebSocket-Extensions: mux; max-channels=4; flow-control, deflate-stream`
This example shows the usage of `mux `extension with parameters `max-channels=4,` `flow-control extension` with no params and `deflate-stream extension` with no parameters.

Reply:
Chrome sends the following header:
`permessage-deflate; client_max_window_bits`
So we would have to split by `,` first and then by `;` right?

Reply:
Yes, that would be the case for parsing with parameters. I currently didn't take parameters into account since I used Java provided classes, but I can parse them if you like.